src/Datastore/DatastoreClient.php

```diff
@@ -0,0 +1,796 @@
+    {
+        return $this->operation->allocateIds($keys, $options);
+    }
+
+    /**
+     * Create a Transaction
+     *
+     * Example:
+     * ```
+     * $transaction = $datastore->transaction();
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/docs/concepts/transactions Datastore Transactions
+     *
+     * @param array $options Configuration options
+     * @return Transaction
+     */
+    public function transaction(array $options = [])
+    {
+        $res = $this->connection->beginTransaction($options + [
+            'projectId' => $this->projectId
+        ]);
+
+        return new Transaction(
+            clone $this->operation,
+            $this->projectId,
+            $res['transaction']
+        );
+    }
+
+    /**
+     * Insert an entity
+     *
+     * An entity with incomplete keys will be allocated an ID prior to insertion.
+     *
+     * Example:
+     * ```
+     * $key = $datastore->key('Person', 'Bob');
+     * $entity = $datastore->entity($key, ['firstName' => 'Bob']);
+     *
+     * $datastore->insert($entity);
+     * ```
+     *
+     * @param Entity $entity The entity to be inserted
+     * @param array $options Configuration options
+     * @return string The entity version
+     * @throws DomainException If a conflict occurs, fail.
+     */
+    public function insert(Entity $entity, array $options = [])
+    {
+        $res = $this->insertBatch([$entity], $options);
+        return $this->parseSingleMutationResult($res);
+    }
+
+    /**
+     * Insert multiple entities
+     *
+     * Any entity with incomplete keys will be allocated an ID prior to insertion.
+     *
+     * Example:
+     * ```
+     * $keys = $datastore->keys(['kind' => 'Person'], [
+     *     'number' => 2
+     * ]);
+     *
+     * $entities = [
+     *     $datastore->entity($key[0], ['firstName' => 'Bob']),
+     *     $datastore->entity($key[1], ['firstName' => 'John'])
+     * ];
+     *
+     * $datastore->insertBatch($entities);
+     * ```
+     *
+     * @param Entity[] $entities The entities to be inserted
+     * @param array $options Configuration options
+     * @return array [Response Body](https://cloud.google.com/datastore/reference/rest/v1/projects/commit#response-body)
+     */
+    public function insertBatch(array $entities, array $options = [])
+    {
+        $entities = $this->operation->allocateIdsToEntities($entities);
+        $this->operation->mutate('insert', $entities, Entity::class);
+        return $this->operation->commit($options);
+    }
+
+    /**
+     * Update an entity
+     *
+     * Please note that updating a record in Cloud Datastore will replace the
+     * existing record. Adding, editing or removing a single property is only
+     * possible by first retrieving the entire entity in its existing state.
+     *
+     * Example:
+     * ```
+     * $entity = $datastore->lookup($ds->key('Person', 'Bob'));
```

Comment:
Seems we would want to be using the same instance of `$datastore` in both of these calls :)